src/odoc/bin/main.ml

```diff
@@ -403,46 +403,56 @@ module Source_tree = struct
 end
 
 module Indexing = struct
-  let output_file ~dst =
-    match dst with
-    | Some file -> Fs.File.of_string file
-    | None -> Fs.File.of_string "index.json"
-
-  let index dst warnings_options inputs_in_file inputs =
-    let output = output_file ~dst in
+  let output_file ~dst marshall =
+    match (dst, marshall) with
+    | Some file, _ -> Fs.File.of_string file
+    | None, `JSON -> Fs.File.of_string "index.json"
+    | None, `Marshall -> Fs.File.of_string "index-index.odoc"
+
+  let index dst marshall warnings_options inputs_in_file inputs =
+    let marshall = if marshall then `Marshall else `JSON in
+    let output = output_file ~dst marshall in
     match (inputs_in_file, inputs) with
     | [], [] ->
         Result.Error
           (`Msg
             "At least one of --file-list or an .odocl file must be passed to \
              odoc compile-index")
-    | _ -> Indexing.compile ~output ~warnings_options inputs_in_file inputs
+    | _ ->
+        Indexing.compile marshall ~output ~warnings_options inputs_in_file
+          inputs
 
   let cmd =
     let dst =
       let doc =
         "Output file path. Non-existing intermediate directories are created. \
-         Defaults to index.json"
+         Defaults to index.json, or index-index.odoc if --marshall is passed \
+         (in which case, the $(i,index-) prefix is mandatory)."
       in
       Arg.(
         value & opt (some string) None & info ~docs ~docv:"PATH" ~doc [ "o" ])
     in
     let inputs_in_file =
       let doc =
-        "Input text file containing a line-separated list of paths to .odocl \
-         files to index."
+        "Input text file containing a line-separated list of paths to \
+         .odocl/.json files to index."
       in
       Arg.(
         value & opt_all convert_fpath []
         & info ~doc ~docv:"FILE" [ "file-list" ])
     in
+    let marshall =
+      let doc = "whether to output a json file, or an .odoc file" in
```

Comment:
Small nitpick : I think it should be "a .odoc file" (pronounced "a dot odoc file")